### PR TITLE
[JENKINS-60564] Rebase before push

### PR DESCRIPTION
**Pull request by johannespfeiffer**
## [JENKINS-60564](https://issues.jenkins-ci.org/browse/JENKINS-60564) Rebase before pushing branches to avoid push rejections because of external repo changes

Assume _during_ our build someone else pushed unrelated changes (commitFile1) to the remote repo. So our own changes committed during the build (commitFile2) cannot be pushed back to the remote origin.

```
* 0eb2599 (HEAD) changed commitFile2
| * 64e71e7 (origin/master) changed commitFile1
|/
* b2578eb init
```

To resolve this we can fetch the remote changes and rebase our own changes just before the push of the branches.

```
* 0e7674c (HEAD) changed commitFile2
* 64e71e7 (origin/master) changed commitFile1
* b2578eb init
```

This PR introduces a new checkbox "Rebase before push" into the "Branches to push to remote repositories". When enabled we try to rebase before the actual push.

## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/jenkinsci/git-plugin/blob/master/CONTRIBUTING.md) doc
- [ ] I have referenced the Jira issue related to my changes in one or more commit messages
- [x] I have added tests that verify my changes
- [x] Unit tests pass locally with my changes
- [x] I have added documentation as necessary
- [x] No Javadoc warnings were introduced with my changes
- [x] No findbugs warnings were introduced with my changes
- [x] I have interactively tested my changes
- [x] Any dependent changes have been merged and published in upstream modules (like git-client-plugin)

## Types of changes

- [x] New feature (non-breaking change which adds functionality)

### CODE REVIEW

**Comment by MarkEWaite** [comment]
I'm quite skeptical of this change because it will push something to the remote repository which is different than what was evaluated by the CI job.  I realize that is your specific intention, but it seems like a use model that I don't think we should encourage.

**Comment by scrwr** [comment]
What we basically try to resolve is, multiple jobs are cloning and pushing to the same repo, usually manipulating only on different files (file per Job). But due to different run-times, they interfere on commits. Rebasing was the most simple solution we could think of and as an optional task, shouldn't be too problematic.

Other solutions might be:
a) add pulling and commiting to the Git-Publisher, to pull and commit near time before before publishing
b) add a pure Pull Task via Git-Plugin, to insert a pull step right before commit and publishing.

Is there any pattern you would prefer over rebasing, or do you have any other recommendation, how to solve our publishing commit conflicts?

**Comment by MarkEWaite** [comment]
My first preference is to have this be added as a separate plugin so that the git plugin does not have another option added to it.  If needed, the git plugin could have one or more extension points added which would allow it to assist with the operation you're performing.

In the context of that additional plugin, I would prefer that it be an extension of the git publisher portion of the plugin rather than being added to the large set of flags and options which control initial checkout.

Unfortunately, I'm describing something that I've never done before.  You'll need to refer to the documentation that describes how to add an extension point, then create a new plugin that implements this additional feature.

**Comment by scrwr** [comment]
The whole option is imho so near to the git-plugin purpose, creating another plugin just for this use case feels like a really bad idea. If there is something, Jenkins really doesn't need, then it is more "one use-case" plugins.

Do you really feel like this option to be such a big alien to the git-plugin? For me it feels like a very typical git approach to pull or rebase before commiting and pushing.

What about solving your second preference only and moving it to the git-publisher portion? Would that be an acceptable compromise?

**Comment by johannespfeiffer** [comment]
> In the context of that additional plugin, I would prefer that it be an extension of the git publisher portion of the plugin rather than being added to the large set of flags and options which control initial checkout.

Just for clarification: That is already the case with the implementation from this PR.

**Comment by johannespfeiffer** [comment]
Hi @MarkEWaite,

regarding your two concerns:

> My first preference is to have this be added as a separate plugin so that the git plugin does not have another option added to it. 

Here is a screenshot where this new and minor option is added with this PR:
ii_jtagsg4w0

So it is not added to this large list of options of the git plugin but rather an (optional) checkbox inside of the git publisher portion.

> [...] it will push something to the remote repository which is different than what was evaluated by the CI job.

I think the most common use case is to have tags for referencing specific revisions. This tags will still stay there and are not affected by the rebase operation, so one can always see what the repo looked like at build time.
So we can see this new rebased revision as an new additional revision in the remote repo (additional to the revision used for the build and marked with the tag). To take my example from the initial post, it would look similar to this:

```
* 0e7674c (HEAD) changed commitFile2
* 64e71e7 (tag: BUILD_1, origin/master) changed commitFile1
| * 0eb2599 (tag: BUILD_2, HEAD) changed commitFile2
|/
* b2578eb init
```

Do you see a way how you would consider integrating this PR?

Thanks
Joe

**Comment by omitrowski** [comment]
Hey guys, this is exactly what I am looking for. My use case is pretty much the same as described above. Several Jenkins jobs updating same repo after a build is completed. Is there any chance for a merge?

**Comment by MarkEWaite** [comment]
@omitrowski there is a good chance this will be merged.  You can help prepare for the merge by downloading the build from the pull request, testing it, and reporting your results on this pull request.

The pull request job generates an [hpi file artifact](https://ci.jenkins.io/job/Plugins/job/git-plugin/job/PR-694/5/artifact/org/jenkins-ci/plugins/git/4.0.0-beta9-rc3094.5a8adf136d1b/git-4.0.0-beta9-rc3094.5a8adf136d1b.hpi) which you can install into your Jenkins server.

**Comment by omitrowski** [comment]
@MarkEWaite I am going to help with the test.

**Comment by omitrowski** [comment]
@MarkEWaite as prommised I performed the test. 

### Full test report
on github: https://github.com/omitrowski/jenkins-git-pugin-test

### Conclusions

 - The patch from the merge request behaves predictable and meets the expected functionality.
 - It does not change the git plugin behavior if option `rebase before push` is not activated.
 - In case of errors the messages are understandable and helpful.
 - As expected in the git log you don't see merges on successful rebase.
-  Endurance test run over 60 times the without issues.

**Comment by MarkEWaite** [comment]
Thanks @omitrowski !

File: src/main/java/hudson/plugins/git/GitPublisher.java [hunk 1]
```diff
@@ -490,15 +500,20 @@ public void setEmptyTargetRepoToOrigin(){
...
         public String getBranchName() {
             return branchName;
         }
+        public boolean isRebaseBeforePush() {
+            return rebaseBeforePush;
+        }
 
         @DataBoundConstructor
-        public BranchToPush(String targetRepoName, String branchName) {
+        public BranchToPush(String targetRepoName, String branchName, boolean rebaseBeforePush) {
```
[the remark this answers is not in the document]
Comment (johannespfeiffer) [reply src/main/java/hudson/plugins/git/GitPublisher.java]: Done.

**Comment by omitrowski** [comment]
@MarkEWaite do you want me to test the latest code status for this request?

**Comment by MarkEWaite** [comment]
@omitrowski we're not yet ready for a second test.  The changes in #689 need to be further tested and merged before this should be tested again.  We also need a test for #693 and #687 and confirmation that the same change which resolves #689 will also resolve those two.

I've delayed merging #689 until I can find the time to create automated tests to verify #693 and #687 and confirm they are fixed.  The automated tests were all passing when 4.0.0-rc was shipped, and all three of those problems escaped into the wild.

**Comment by s4ros** [comment]
I've hit the same issue. Please merge 🙏

**Comment by MarkEWaite** [comment]
@omitrowski #689 has been merged and #693 and #687 are both confirmed by automated tests and by interactive testing that they are resolved.  This is ready to be tested again.

**Comment by omitrowski** [comment]
> @omitrowski #689 has been merged and #693 and #687 are both confirmed by automated tests and by interactive testing that they are resolved. This is ready to be tested again.
@MarkEWaite I successfully tested the merge request using this versions:
[#17 git plugin 4.0.0-beta10-rc3175](https://ci.jenkins.io/job/Plugins/job/git-plugin/job/PR-694/17/artifact/org/jenkins-ci/plugins/git/4.0.0-beta10-rc3175.8a809a74f11b/git-4.0.0-beta10-rc3175.8a809a74f11b.hpi)
[#301 git-client 3.0.0-beta10-rc1957](https://ci.jenkins.io/job/Plugins/job/git-client-plugin/job/master/301/artifact/org/jenkins-ci/plugins/git-client/3.0.0-beta10-rc1957.86da9c9ee817/git-client-3.0.0-beta10-rc1957.86da9c9ee817.hpi)

### Full test report
on github: https://github.com/omitrowski/jenkins-git-pugin-test (take-2)

### Conclusions
The patch from the merge request behaves predictable and meets the expected functionality.
It does not change the git plugin behavior if option rebase before push is not activated.
In case of errors the messages are understandable and helpful.
As expected in the git log you don't see merges on successful rebase.
Endurance test run over 90 times the without issues.

**Comment by omitrowski** [comment]
@MarkEWaite any news from your side about the progress?

**Comment by MarkEWaite** [comment]
No progress.  I've been fully occupied with DevOps World | Jenkins World 2019 preparations and am now at the last day of the conference.  After some recovery time, I'll return to work and to my regularly scheduled weekends and nights.  There should be more time in those upcoming weekends and nights to review this further.

**Comment by leo-chan-symc** [comment]
I tested this pull request and the rebase option works well for me.

**Pull request merged**